### PR TITLE
[improve](jsonb)Invalid json path prompts an error instead of null

### PR DESCRIPTION
**Pull request by LemonLiTree**
1. Invalid json path prompts an error instead of null：
before：
```sql
mysql> SELECT jsonb_extract('[{"k1":"v41","k2":400},1,"a",3.14]', '$[a]');
+-------------------------------------------------------------+
| jsonb_extract('[{"k1":"v41","k2":400},1,"a",3.14]', '$[a]') |
+-------------------------------------------------------------+
| NULL                                                        |
+-------------------------------------------------------------+
1 row in set (0.01 sec)
```
now
```sql
mysql> SELECT jsonb_extract('[{"k1":"v41","k2":400},1,"a",3.14]', '$[a]');
ERROR 1105 (HY000): errCode = 2, detailMessage = (127.0.0.1)[INVALID_ARGUMENT]Json path error: Invalid Json Path for value: $[a]
```
2. fix some problem: https://github.com/apache/doris/pull/19185
   a. support negative numbers
```sql
mysql> SELECT jsonb_extract('[{"k1":"v41","k2":400},1,"a",3.14]', '$[-2]');
+--------------------------------------------------------------+
| jsonb_extract('[{"k1":"v41","k2":400},1,"a",3.14]', '$[-2]') |
+--------------------------------------------------------------+
| "a"                                                          |
+--------------------------------------------------------------+
1 row in set (0.02 sec)
```
  b. Avoid using unnecessary memory
3. Supplementary regression test

### CODE REVIEW

**Comment by LemonLiTree** [comment]
run buildall

**Review by xiaokang — COMMENTED** [review]
rebase master

File: be/src/util/jsonb_document.h [hunk 1]
```diff
@@ -273,35 +276,43 @@ class Stream {
         skip_whitespace();
     }
 
-    void clear_legPtr() { legPtr = nullptr; }
+    void clear_legPtr() { leg_ptr = nullptr; }
```
[the remark this answers is not in the document]
Comment (LemonLiTree) [reply be/src/util/jsonb_document.h]: done

File: be/src/util/jsonb_document.h [hunk 2]
```diff
@@ -273,35 +276,43 @@ class Stream {
...
 
     void set_legPtr(char* ptr) {
```
[the remark this answers is not in the document]
Comment (LemonLiTree) [reply be/src/util/jsonb_document.h]: done

File: be/src/util/jsonb_document.h [hunk 3]
```diff
@@ -273,35 +276,43 @@ class Stream {
...
         clear_legPtr();
-        legPtr = ptr;
+        leg_ptr = ptr;
     }
 
-    char* get_legPtr() { return legPtr; }
+    char* get_legPtr() { return leg_ptr; }
```
[the remark this answers is not in the document]
Comment (LemonLiTree) [reply be/src/util/jsonb_document.h]: done

File: be/src/util/jsonb_document.h [hunk 4]
```diff
@@ -273,35 +276,43 @@ class Stream {
...
 
-    void clear_legLen() { legLen = 0; }
+    void clear_legLen() { leg_len = 0; }
```
[the remark this answers is not in the document]
Comment (LemonLiTree) [reply be/src/util/jsonb_document.h]: done

File: be/src/util/jsonb_document.h [hunk 5]
```diff
@@ -273,35 +276,43 @@ class Stream {
...
 
-    void add_legLen() { legLen++; }
+    void add_legLen() { leg_len++; }
```
[the remark this answers is not in the document]
Comment (LemonLiTree) [reply be/src/util/jsonb_document.h]: done

File: be/src/util/jsonb_document.h [hunk 6]
```diff
@@ -273,35 +276,43 @@ class Stream {
...
 
-    unsigned int get_legLen() { return legLen; }
+    unsigned int get_legLen() const { return leg_len; }
```
[the remark this answers is not in the document]
Comment (LemonLiTree) [reply be/src/util/jsonb_document.h]: done

File: be/src/util/jsonb_document.h [hunk 7]
```diff
@@ -273,35 +276,43 @@ class Stream {
...
             }
         }
-        legPtr[new_len] = '\0';
-        legLen = new_len;
+        leg_ptr[new_len] = '\0';
+        leg_len = new_len;
     }
 
-    void set_hasEscapes(bool has) { hasEscapes = has; }
+    void set_hasEscapes(bool has) { has_escapes = has; }
```
[the remark this answers is not in the document]
Comment (LemonLiTree) [reply be/src/util/jsonb_document.h]: done

File: be/src/util/jsonb_document.h [hunk 8]
```diff
@@ -273,35 +276,43 @@ class Stream {
...
 
-    bool get_hasEscapes() { return hasEscapes; }
+    bool get_hasEscapes() const { return has_escapes; }
```
[the remark this answers is not in the document]
Comment (LemonLiTree) [reply be/src/util/jsonb_document.h]: done

File: be/src/util/jsonb_document.h [hunk 9]
```diff
@@ -1200,52 +1225,125 @@ inline JsonbValue* JsonbValue::findPath(const char* key_path, unsigned int kp_le
...
         }
 
-        if (LIKELY(pval->type_ == JsonbType::T_Object)) {
-            if (stream.get_legLen() == 1 && *stream.get_legPtr() == WILDCARD) {
-                return pval;
-            } else if (stream.get_hasEscapes()) {
-                stream.remove_escapes();
-            }
+        if (stream.get_type() == MEMBER_CODE) {
+            if (LIKELY(pval->type_ == JsonbType::T_Array)) {
```
[the remark this answers is not in the document]
Comment (LemonLiTree) [reply be/src/util/jsonb_document.h]: done

File: be/src/util/jsonb_document.h [hunk 10]
```diff
@@ -1200,52 +1225,125 @@ inline JsonbValue* JsonbValue::findPath(const char* key_path, unsigned int kp_le
...
+                return nullptr;
+            } else if (LIKELY(pval->type_ == JsonbType::T_Object)) {
+                if (stream.get_legLen() == 1 && *stream.get_legPtr() == WILDCARD) {
+                    continue;
```
[the remark this answers is not in the document]
Comment (LemonLiTree) [reply be/src/util/jsonb_document.h]: done

File: be/src/util/jsonb_document.h [hunk 11]
```diff
@@ -1200,52 +1225,125 @@ inline JsonbValue* JsonbValue::findPath(const char* key_path, unsigned int kp_le
...
+        } else if (stream.get_type() == ARRAY_CODE) {
             int index = 0;
-            std::string idx_string(stream.get_legPtr(), stream.get_legLen());
+            std::string_view idx_string(stream.get_legPtr(), stream.get_legLen());
 
             if (stream.get_legLen() == 1 && *stream.get_legPtr() == WILDCARD) {
-                return pval;
-            } else if (std::string(stream.get_legPtr(), 4) == LAST) {
+                if (LIKELY(pval->type_ == JsonbType::T_Object)) {
+                    return nullptr;
```
[the remark this answers is not in the document]
Comment (LemonLiTree) [reply be/src/util/jsonb_document.h]: done

File: be/src/util/jsonb_document.h [hunk 12]
```diff
@@ -1200,52 +1225,125 @@ inline JsonbValue* JsonbValue::findPath(const char* key_path, unsigned int kp_le
...
+                } else if (LIKELY(pval->type_ == JsonbType::T_Array)) {
+                    continue;
```
[the remark this answers is not in the document]
Comment (LemonLiTree) [reply be/src/util/jsonb_document.h:1260]: In order to support '$[*][0]', but should skip(1) and skip_whitespace() before continue

File: be/src/util/jsonb_document.h [hunk 13]
```diff
@@ -1200,52 +1225,125 @@ inline JsonbValue* JsonbValue::findPath(const char* key_path, unsigned int kp_le
...
+                auto result = std::from_chars(idx_string.data(),
+                                              idx_string.data() + idx_string.size(), index);
+                if (result.ec != std::errc()) {
+                    is_invalid_json_path = true;
+                    return nullptr;
+                }
+
+                if (LIKELY(pval->type_ == JsonbType::T_Object)) {
+                    if (index == 0) {
+                        continue;
```
[the remark this answers is not in the document]
Comment (LemonLiTree) [reply be/src/util/jsonb_document.h:1317]: yeah, mysql supports such

File: be/src/util/jsonb_document.h [hunk 14]
```diff
@@ -1200,52 +1225,125 @@ inline JsonbValue* JsonbValue::findPath(const char* key_path, unsigned int kp_le
...
-                return pval;
-            } else if (std::string(stream.get_legPtr(), 4) == LAST) {
+                if (LIKELY(pval->type_ == JsonbType::T_Object)) {
+                    return nullptr;
+                } else if (LIKELY(pval->type_ == JsonbType::T_Array)) {
+                    continue;
+                } else {
+                    return nullptr;
+                }
+            } else if (*stream.get_legPtr() == MINUS && stream.get_legLen() > 1) {
```
[the remark this answers is not in the document]
Comment (LemonLiTree) [reply be/src/util/jsonb_document.h]: done

**Comment by LemonLiTree** [comment]
run buildall

**Comment by LemonLiTree** [comment]
run buildall

**Comment by LemonLiTree** [comment]
run buildall

**Comment by LemonLiTree** [comment]
run buildall

**Comment by LemonLiTree** [comment]
run buildall

**Review by xiaokang — APPROVED** [review]
LGTM

**Review by zy-kkk — APPROVED** [review]

**Pull request merged**